src/main/scala/com/fulcrumgenomics/bam/Pileup.scala

```diff
@@ -0,0 +1,209 @@
+  requireValidOffset()
+
+}
+
+/** Pileup entry representing a deletion. */
+case class DeletionEntry(override val rec: SAMRecord, override val offset: Int) extends PileupEntry {
+  requireValidOffset()
+}
+
+
+/**
+  * Class that provides methods to build and filter [[Pileup]]s.
+  *
+  * @param dict the SAMSequenceDictionary for the reference sequence
+  * @param minMapQ the minimum MAPQ to allow a read to contribute to the pileup
+  * @param minBaseQ the minimum base quality to allow a base to contribute to the pileup
+  * @param mappedPairsOnly if true only allow reads from read-pairs with both reads mapped to contribute to the pileup
+  * @param includeDuplicates if true allow reads flagged as duplicates to contribute to the pileup
+  * @param includeSecondaryAlignments if true allow secondary alignments to contribute to the pileup
+  * @param includeSupplementalAlignments if true allow supplemental alignments to contribute to the pileup
+  */
+class PileupBuilder(dict: SAMSequenceDictionary,
+                    minMapQ: Int = 20,
+                    minBaseQ: Int = 20,
+                    mappedPairsOnly: Boolean = true,
+                    includeDuplicates: Boolean = false,
+                    includeSecondaryAlignments: Boolean = false,
+                    includeSupplementalAlignments: Boolean = false) {
+
+  private val readFilters = new ListBuffer[SAMRecord => Boolean]()
+  private val baseFilters = new ListBuffer[PileupEntry => Boolean]()
+
+  // Setup the default read level filters
+  if (minMapQ > 0)                    readFilters += { r => r.getMappingQuality >= minMapQ }
+  if (mappedPairsOnly)                readFilters += { r => r.getReadPairedFlag && !r.getReadUnmappedFlag && !r.getMateUnmappedFlag }
+  if (!includeDuplicates)             readFilters += { r => !r.getDuplicateReadFlag }
+  if (!includeSecondaryAlignments)    readFilters += { r => !r.getNotPrimaryAlignmentFlag }
+  if (!includeSupplementalAlignments) readFilters += { r => !r.getSupplementaryAlignmentFlag }
+
+  // Setup the default base level filters
+  if (minBaseQ > 0) baseFilters += {
+    case p: BaseEntry => p.qual >= minBaseQ
+    case _ => true
+  }
+
+  /** Adds a read level filter to the set of filters. */
+  def withReadFilter(f: SAMRecord => Boolean): PileupBuilder = yieldAndThen(this) { this.readFilters += f }
+
+  /** Adds a base level filter to the set of filters. */
+  def withBaseFilter(f: PileupEntry => Boolean): PileupBuilder = yieldAndThen(this) { this.baseFilters += f }
+
+  /** Checks to see if all read filters accept the read. */
+  protected def accepts(rec: SAMRecord): Boolean = this.readFilters.forall(f => f(rec))
+
+  /** Checks to see if all read filters accept the read. */
+  protected def accepts(entry: PileupEntry): Boolean = this.baseFilters.forall(f => f(entry))
+
+  /**
+    * Constructs a pileup of a single position from the records returned by the iterator.
```

Comment:
When I see "a single position" I think which one?  The position is given as a parameter, so perhaps a compromise would be "the given position" instead.